### ISSUE TITLE
Stray record field assignments are allowed as expressions

### ISSUE THREAD

**Issue by vpax** [zeek/zeek#2763]
The following code:
```
print $foo=5;
```
surprisingly just goes ahead and prints `5`.  (This is abstracted from a nastier instance where a record was getting mis-constructed because one of its fields was mistakenly appearing inside an embedded function call, rather than after it.)

**Comment by timwoj** [comment]
This happens because there's zero validation that the field name there is a valid ID: https://github.com/zeek/zeek/blob/7f96fc08fb971d61099bde4dbfa9c74e7f2e4077/src/parse.y#L770-L774

We create a `FieldAssignExpr` off the bare name and when the `PrintStmt` is called, the evaluation of the underlying `UnaryExpr` just returns the value to be printed.

**Comment by vpax** [comment]
Yeah.  I _believe_ that these expressions are only valid inside of record constructors.  With luck that can be addressed by changing these `expr` productions in the grammar:
```
|       '[' opt_expr_list ']'
...
|       TOK_RECORD '(' expr_list ')'
```
to use different notions of expression lists, ones that allow their elements to have $field= elements, and not allowing such expressions elsewhere in the grammar.  I can take a swing at this (though have plenty of other stuff on my plate, so it'll take a while ...)
